File: app/src/main/java/com/savvasdalkitsis/gameframe/widget/presenter/WidgetPresenter.java
package com.savvasdalkitsis.gameframe.widget.presenter;

import android.util.Log;

import com.savvasdalkitsis.gameframe.control.view.PowerTileService;
import com.savvasdalkitsis.gameframe.infra.navigation.Navigator;
import com.savvasdalkitsis.gameframe.ip.repository.IpRepository;
import com.savvasdalkitsis.gameframe.rx.RxTransformers;
import com.savvasdalkitsis.gameframe.gameframe.usecase.GameFrameUseCase;
import com.savvasdalkitsis.gameframe.widget.view.WidgetView;

import rx.Observable;

public class WidgetPresenter {

    private final GameFrameUseCase gameFrameUseCase;
    private final IpRepository ipRepository;
    private final Navigator navigator;
    private WidgetView widgetView;

    public WidgetPresenter(GameFrameUseCase gameFrameUseCase, IpRepository ipRepository, Navigator navigator) {
        this.gameFrameUseCase = gameFrameUseCase;
        this.ipRepository = ipRepository;
        this.navigator = navigator;
    }

    public void bindView(WidgetView widgetView) {
        this.widgetView = widgetView;
    }

    public void menu() {
        perform(gameFrameUseCase.menu());
    }

    public void next() {
        perform(gameFrameUseCase.next());
    }

    public void power() {
        perform(gameFrameUseCase.togglePower());
    }

    private void perform(Observable<Void> operation) {
        ipRepository.getIpAddress()
                .doOnError(e -> {
                    Log.e(PowerTileService.class.getName(), "IP address not found", e);
                    navigator.navigateToIpSetup();
                })
                .onErrorResumeNext(Observable.empty())
                .flatMap(ip -> operation)
                .compose(RxTransformers.schedulers())
                .subscribe(n -> {}, e -> {
                    Log.e(PowerTileService.class.getName(), "Error communicating with the GameFrame", e);
                    widgetView.operationError();
                });
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/widget/presenter/WidgetPresenter.kt
package com.savvasdalkitsis.gameframe.widget.presenter

import android.util.Log
import com.savvasdalkitsis.gameframe.control.view.PowerTileService
import com.savvasdalkitsis.gameframe.gameframe.usecase.GameFrameUseCase
import com.savvasdalkitsis.gameframe.navigation.Navigator
import com.savvasdalkitsis.gameframe.ip.repository.IpRepository
import com.savvasdalkitsis.gameframe.infra.rx.RxTransformers
import com.savvasdalkitsis.gameframe.widget.view.WidgetView
import io.reactivex.Completable

class WidgetPresenter(private val gameFrameUseCase: GameFrameUseCase,
                      private val ipRepository: IpRepository,
                      private val navigator: Navigator) {

    private var widgetView: WidgetView? = null

    fun bindView(widgetView: WidgetView) {
        this.widgetView = widgetView
    }

    fun menu() = perform(gameFrameUseCase.menu())

    fun next() = perform(gameFrameUseCase.next())

    fun power() = perform(gameFrameUseCase.togglePower())

    private fun perform(operation: Completable) {
        ipRepository.ipAddress
                .doOnError {
                    Log.e(PowerTileService::class.java.name, "IP address not found", it)
                    navigator.navigateToIpSetup()
                }
                .flatMapCompletable { operation }
                .compose(RxTransformers.schedulers())
                .subscribe({ }, {
                    Log.e(PowerTileService::class.java.name, "Error communicating with the GameFrame", it)
                    widgetView?.operationError()
                })
    }
}
